### PR TITLE
Distributor: Add limits for Otel write request byte size

### PR DESCRIPTION
**Pull request by ying-jeanne**
#### Which issue(s) this PR fixes or relates to

The current distributor message limit is 100MB See [doc](https://grafana.com/docs/mimir/latest/configure/configuration-parameters/#server), Which demonstrates too much for compressed Otel write request. This could cause problem for otel, end up in big decompressed message and timeout. Knowing that otlp collector could have different compression algorithm and also can also be configured to be no compression. See [doc](https://github.com/open-telemetry/opentelemetry-collector/tree/main/config/confighttp), introduce new uncompressed request byte size in order to reject abusive write request as early as possible.

**Why not using existing `distributor.max-recv-msg-size`?**
- The remote write and otlp request doesn't write request with same format, doesn't contains same quantity samples when requests are the same size. 
- In the same container, we can have several users who use otel and remote write during the same time, they need to be limited with different values

**About default value**
- The current big request is rejected by 413 error code, which is not retryable, just to be on the safe side, put 100M as limit, we saw 100MiB request end up executing 10s. This needs to be carefully tested/observed in order to find the final sweet default value.

**Adding runbook dedicated to the error**

#### Checklist

- [x] Tests updated.
- [x] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [x] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/distributor/distributor.go [hunk 1]
```diff
@@ -236,7 +237,8 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet, logger log.Logger) {
 	cfg.DistributorRing.RegisterFlags(f, logger)
 	cfg.RetryConfig.RegisterFlags(f)
 
-	f.IntVar(&cfg.MaxRecvMsgSize, "distributor.max-recv-msg-size", 100<<20, "Max message size in bytes that the distributors will accept for incoming push requests to the remote write API. If exceeded, the request will be rejected.")
+	f.IntVar(&cfg.MaxRecvMsgSize, "distributor.max-recv-msg-size", 80<<20, "Max message size in bytes that the distributors will accept for incoming push requests to the remote write API. If exceeded, the request will be rejected.")
```
Comment (pstibrany) [inline pkg/distributor/distributor.go:240]: Why was this default value changed? That seems unrelated to this PR.

File: pkg/distributor/distributor.go [hunk 2]
```diff
@@ -185,9 +185,10 @@ type Config struct {
 	RetryConfig     RetryConfig     `yaml:"retry_after_header"`
 	HATrackerConfig HATrackerConfig `yaml:"ha_tracker"`
 
-	MaxRecvMsgSize           int           `yaml:"max_recv_msg_size" category:"advanced"`
-	MaxRequestPoolBufferSize int           `yaml:"max_request_pool_buffer_size" category:"experimental"`
-	RemoteTimeout            time.Duration `yaml:"remote_timeout" category:"advanced"`
+	MaxRecvMsgSize                 int           `yaml:"max_recv_msg_size" category:"advanced"`
+	MaxOtelDecompressedRecvMsgSize int           `yaml:"max_otel_decompressed_recv_msg_size" category:"experimental"`
```
Comment (pstibrany) [inline pkg/distributor/distributor.go:189]: I would suggest to use the correct naming here:

```suggestion
	MaxOtlpRequestSize int           `yaml:"max_otlp_request_size" category:"experimental"`
```

Since Mimir accepts "OTLP/HTTP requests".

The fact that it can be compressed is irrelevant, and it doesn't make much sense to apply limit to compressed message only, so I dropped "Decompressed" part of the name (everywhere, not just here).

File: pkg/distributor/distributor.go [hunk 3]
```diff
@@ -236,7 +237,8 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet, logger log.Logger) {
 	cfg.DistributorRing.RegisterFlags(f, logger)
 	cfg.RetryConfig.RegisterFlags(f)
 
-	f.IntVar(&cfg.MaxRecvMsgSize, "distributor.max-recv-msg-size", 100<<20, "Max message size in bytes that the distributors will accept for incoming push requests to the remote write API. If exceeded, the request will be rejected.")
+	f.IntVar(&cfg.MaxRecvMsgSize, "distributor.max-recv-msg-size", 80<<20, "Max message size in bytes that the distributors will accept for incoming push requests to the remote write API. If exceeded, the request will be rejected.")
+	f.IntVar(&cfg.MaxOtelDecompressedRecvMsgSize, "distributor.max-otel-decompressed-recv-msg-size", 100<<20, "Maximum message size in bytes that the distributors will accept for incoming decompressed write requests to the OTEL API. Requests exceeding this limit will be rejected.")
```
Comment (pstibrany) [inline pkg/distributor/distributor.go:241]:
```suggestion
	f.IntVar(&cfg.MaxOtlpRequestSize, "distributor.max-otlp-request-size", 100<<20, "Maximum OTLP request size in bytes that the distributors will accept. Requests exceeding this limit will be rejected.")
```

File: pkg/distributor/otel.go [hunk 4]
```diff
@@ -53,7 +54,7 @@ type OTLPHandlerLimits interface {
 
 // OTLPHandler is an http.Handler accepting OTLP write requests.
 func OTLPHandler(
-	maxRecvMsgSize int,
+	maxOtelDecompressedRecvMsgSize int,
```
Comment (pstibrany) [inline pkg/distributor/otel.go:57]: I don't see good reason to rename this parameter, and it increases diff.

"Otel" part is unnecessary, since we're already inside OTLPHandler

"decompressed" is implied -- it doesn't make sense to apply limit to compressed size only.

If we need to rename this, then I'd suggest "maxRequestSize", since we're dealing with "OTLP requests".

File: pkg/distributor/otel.go [hunk 5]
```diff
@@ -209,7 +215,7 @@ func OTLPHandler(
 }
 
 func otlpHandler(
-	maxRecvMsgSize int,
+	maxOtelDecompressedRecvMsgSize int,
```
Comment (pstibrany) [inline pkg/distributor/otel.go:218]: Same comment as in `OTLPHandler` applies here.

File: pkg/distributor/otel_test.go [hunk 6]
```diff
@@ -245,9 +245,9 @@ func TestHandlerOTLPPush(t *testing.T) {
 		series   []prompb.TimeSeries
 		metadata []mimirpb.MetricMetadata
 
-		compression bool
-		encoding    string
-		maxMsgSize  int
+		compression            bool
+		encoding               string
+		maxDecompressedMsgSize int
```
Comment (pstibrany) [inline pkg/distributor/otel_test.go:250]: Unnecessary change.

File: pkg/distributor/push.go [hunk 7]
```diff
@@ -100,14 +100,22 @@ func Handler(
...
 
 func (e distributorMaxWriteMessageSizeErr) Error() string {
 	msgSizeDesc := fmt.Sprintf(" of %d bytes", e.actual)
+	var flag string
 	if e.actual < 0 {
 		msgSizeDesc = ""
 	}
-	return globalerror.DistributorMaxWriteMessageSize.MessageWithPerInstanceLimitConfig(fmt.Sprintf("the incoming push request has been rejected because its message size%s is larger than the allowed limit of %d bytes", msgSizeDesc, e.limit), "distributor.max-recv-msg-size")
+	if e.id == globalerror.DistributorMaxOtelDecompressedWriteMessageSize {
+		flag = "distributor.max_otel_decompressed_recv_msg_size"
```
Comment (pstibrany) [inline pkg/distributor/push.go:113]: `-distributor.max_otel_decompressed_recv_msg_size` is incorrect. To avoid mismatch, extract the flag name to a string constant and use it in all places where needed.

Same comment about extracting flag name to constant applies to `distributor.max-recv-msg-size`.

File: pkg/distributor/push.go [hunk 8]
```diff
@@ -100,14 +100,22 @@ func Handler(
...
 }
 
 func (e distributorMaxWriteMessageSizeErr) Error() string {
 	msgSizeDesc := fmt.Sprintf(" of %d bytes", e.actual)
+	var flag string
 	if e.actual < 0 {
 		msgSizeDesc = ""
 	}
-	return globalerror.DistributorMaxWriteMessageSize.MessageWithPerInstanceLimitConfig(fmt.Sprintf("the incoming push request has been rejected because its message size%s is larger than the allowed limit of %d bytes", msgSizeDesc, e.limit), "distributor.max-recv-msg-size")
+	if e.id == globalerror.DistributorMaxOtelDecompressedWriteMessageSize {
```
Comment (pstibrany) [inline pkg/distributor/push.go:112]: Instead of having extra logic in `Error` method, can we have two separate types for each error (one for OTLP, one for Push) or put `flag` into existing type and create two separate constructors (`newPushMaxMessageSizeError` and `newOTLPMaxMessageSizeError`) for each type.

File: pkg/util/globalerror/user.go [hunk 9]
```diff
@@ -74,8 +74,9 @@ const (
 	StoreConsistencyCheckFailed ID = "store-consistency-check-failed"
 	BucketIndexTooOld           ID = "bucket-index-too-old"
 
-	DistributorMaxWriteMessageSize         ID = "distributor-max-write-message-size"
-	DistributorMaxWriteRequestDataItemSize ID = "distributor-max-write-request-data-item-size"
+	DistributorMaxWriteMessageSize                 ID = "distributor-max-write-message-size"
+	DistributorMaxOtelDecompressedWriteMessageSize ID = "distributor-max-otel-decompressed-write-message-size"
```
Comment (pstibrany) [inline pkg/util/globalerror/user.go:78]:
```suggestion
	DistributorMaxOtlpRequestSize ID = "distributor-max-otlp-request-size"
```

**Review by pstibrany — COMMENTED** [review]
Thank you for working on this.

File: CHANGELOG.md [hunk 10]
```diff
@@ -19,6 +19,7 @@
     * `-ingest-storage.read-consistency`: configures the default read consistency.
     * `-ingest-storage.migration.distributor-send-to-ingesters-enabled`: enabled tee-ing writes to classic ingesters and Kafka, used during a live migration to the new ingest storage architecture.
     * `-ingester.partition-ring.*`: configures partitions ring backend.
+* [ENHANCEMENT] Distributor: Add `max_otel_decompressed_recv_msg_size` to limit the otel decompressed write request byte size. #8574
```
Comment (aknuds1) [inline CHANGELOG.md:22]: [Nit] Don't we normally refer to flags rather than YAML parameters in the changelog? For example `-ingest-storage.enabled` just above this changelog entry.

Also, I think the flag should refer to OTLP rather than OTel, since it controls the OTLP endpoint specifically.

```suggestion
* [ENHANCEMENT] Distributor: Add configuration option `-distributor.max-otlp-decompressed-recv-msg-size` to limit otel decompressed write request byte sizes. #8574
```

File: docs/sources/mimir/configure/about-versioning.md [hunk 11]
```diff
@@ -74,6 +74,8 @@ The following features are currently experimental:
     - `-distributor.retry-after-header.max-backoff-exponent`
   - Limit exemplars per series per request
     - `-distributor.max-exemplars-per-series-per-request`
+  - Limit otel write request byte size
```
Comment (aknuds1) [inline docs/sources/mimir/configure/about-versioning.md:77]: I think OTLP is more specific:

```suggestion
  - Limit OTLP write request byte size
```

File: pkg/distributor/distributor.go (hunk 2 above)
Comment (aknuds1) [reply pkg/distributor/distributor.go:189]: I agree it should be OTLP, rather than OTel. Style nit, can we write `OTLP` here to be consistent with Go's consistently capitalized initialisms idiom?

File: pkg/distributor/distributor.go (hunk 1 above)
Comment (aknuds1) [reply pkg/distributor/distributor.go:240]: Yep, I was going to ask the same as pstibrany.

File: pkg/distributor/otel.go (hunk 4 above)
Comment (aknuds1) [reply pkg/distributor/otel.go:57]: Agree there's no point in renaming this parameter. It's implied that it's OTLP.

```suggestion
	maxRecvMsgSize int,
```

File: pkg/distributor/otel_test.go (hunk 6 above)
Comment (aknuds1) [reply pkg/distributor/otel_test.go:250]: Agree w/ @pstibrany.

```suggestion
		maxMsgSize int
```

File: pkg/distributor/otel_test.go [hunk 12]
```diff
@@ -278,33 +278,33 @@ func TestHandlerOTLPPush(t *testing.T) {
...
 			responseCode:              http.StatusOK,
 			enableOtelMetadataStorage: true,
 		},
 		{
-			name:        "Write samples. Request too big",
-			compression: false,
-			maxMsgSize:  30,
-			series:      sampleSeries,
-			metadata:    sampleMetadata,
+			name:                   "Write samples. No compression, Request too big",
```
Comment (aknuds1) [inline pkg/distributor/otel_test.go:289]:
```suggestion
			name:                   "Write samples. No compression, request too big",
```

File: pkg/distributor/otel_test.go [hunk 13]
```diff
@@ -314,10 +314,24 @@ func TestHandlerOTLPPush(t *testing.T) {
 			expectedLogs: []string{`level=error user=test msg="detected an error while ingesting OTLP metrics request (the request may have been partially ingested)" httpCode=415 err="rpc error: code = Code(415) desc = unsupported compression: snappy. Only \"gzip\" or no compression supported" insight=true`},
 		},
 		{
-			name:       "Rate limited request",
-			maxMsgSize: 100000,
-			series:     sampleSeries,
-			metadata:   sampleMetadata,
+			name:                   "Write samples. With compression, Request too big",
```
Comment (aknuds1) [inline pkg/distributor/otel_test.go:317]:
```suggestion
			name:                   "Write samples. With compression, request too big",
```

File: docs/sources/mimir/manage/mimir-runbooks/_index.md [hunk 14]
```diff
@@ -2236,6 +2236,20 @@ How to **fix** it:
...
 - Increase the allowed limit by using the `-distributor.max-recv-msg-size` option.
 
+### err-mimir-distributor-max-otel-decompressed-write-message-size
+
+This error occurs when a distributor rejects an OTEL write request because its message size is larger than the allowed limit after decompression.
+
+How it **works**:
+
+- The distributor implements an upper limit on the message size of incoming otel write requests after decompression regardless compression type. See more otlp collector [compression details](https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/configgrpc/README.md#client-configuration).
+- To configure the limit, set the `-distributor.max_otel_decompressed_recv_msg_size`
```
Comment (aknuds1) [inline docs/sources/mimir/manage/mimir-runbooks/_index.md:2246]: Clearly flag must be this:
```suggestion
- To configure the limit, set the `-distributor.max-otel-decompressed-recv_msg-size`
```

File: docs/sources/mimir/manage/mimir-runbooks/_index.md [hunk 15]
```diff
@@ -2236,6 +2236,20 @@ How to **fix** it:
...
+
+How to **fix** it:
+
+- Increase the allowed limit by using the `-distributor.max_otel_decompressed_recv_msg_size` option.
```
Comment (aknuds1) [inline docs/sources/mimir/manage/mimir-runbooks/_index.md:2250]: Same mistyping as above:

File: pkg/distributor/push.go (hunk 8 above)
Comment (aknuds1) [reply pkg/distributor/push.go:112]: Agree :point_up:

**Review by aknuds1 — CHANGES_REQUESTED** [review]
Please see comments.

File: pkg/distributor/otel.go (hunk 4 above)
Comment (ying-jeanne) [reply pkg/distributor/otel.go:57]: the decompressed is to clarify that we limit on request bytes size after they decompress. I am ok to rename it, just found distributor.recv-msg-size unclear whether it is checking before decompression or after and had to look into code to make sure.

File: pkg/distributor/push.go (hunk 8 above)
Comment (ying-jeanne) [reply pkg/distributor/push.go:112]: was thinking about those, didn't want to duplicate the code, if we find this more clear, I am open to change.

File: pkg/distributor/otel.go (hunk 4 above)
Comment (aknuds1) [reply pkg/distributor/otel.go:57]: I still think it's unnecessary to include `Decompressed` in the arg name. The original name is fine.

File: docs/sources/mimir/manage/mimir-runbooks/_index.md [hunk 16]
```diff
@@ -2236,6 +2236,20 @@ How to **fix** it:
 
 - Increase the allowed limit by using the `-distributor.max-recv-msg-size` option.
 
+### err-mimir-distributor-max-otel-decompressed-write-message-size
+
+This error occurs when a distributor rejects an OTEL write request because its message size is larger than the allowed limit after decompression.
```
Comment (tacole02) [inline docs/sources/mimir/manage/mimir-runbooks/_index.md:2241]:
```suggestion
This error occurs when a distributor rejects an OTel write request because its message size is larger than the allowed limit after decompression.
```

File: pkg/distributor/otel.go (hunk 4 above)
Comment (ying-jeanne) [reply pkg/distributor/otel.go]: I have no strong opinion on naming here, renaming is done in commit 1ba1bcb4854d016dfada97077fc1696ea23c84e3

File: pkg/distributor/otel.go (hunk 5 above)
Comment (ying-jeanne) [reply pkg/distributor/otel.go]: renaming is done in commit https://github.com/grafana/mimir/commit/1ba1bcb4854d016dfada97077fc1696ea23c84e3

File: pkg/distributor/otel_test.go (hunk 6 above)
Comment (ying-jeanne) [reply pkg/distributor/otel_test.go]: renaming is done in commit https://github.com/grafana/mimir/commit/1ba1bcb4854d016dfada97077fc1696ea23c84e3

File: pkg/distributor/push.go (hunk 7 above)
Comment (ying-jeanne) [reply pkg/distributor/push.go]: renaming and const is done in commit https://github.com/grafana/mimir/commit/1ba1bcb4854d016dfada97077fc1696ea23c84e3

File: pkg/distributor/push.go (hunk 8 above)
Comment (ying-jeanne) [reply pkg/distributor/push.go]: error split is done in commit https://github.com/grafana/mimir/commit/1ba1bcb4854d016dfada97077fc1696ea23c84e3

File: pkg/util/globalerror/user.go (hunk 9 above)
Comment (ying-jeanne) [reply pkg/util/globalerror/user.go]: renaming is done in commit https://github.com/grafana/mimir/commit/1ba1bcb4854d016dfada97077fc1696ea23c84e3

File: docs/sources/mimir/manage/mimir-runbooks/_index.md (hunk 14 above)
Comment (ying-jeanne) [reply docs/sources/mimir/manage/mimir-runbooks/_index.md]: flag updated in commit https://github.com/grafana/mimir/commit/1ba1bcb4854d016dfada97077fc1696ea23c84e3

File: pkg/distributor/distributor.go (hunk 1 above)
Comment (ying-jeanne) [reply pkg/distributor/distributor.go]: correct in 1ba1bcb4854d016dfada97077fc1696ea23c84e3, the 80M is for otlp instead.

File: pkg/distributor/distributor.go (hunk 2 above)
Comment (ying-jeanne) [reply pkg/distributor/distributor.go]: rename done in 1ba1bcb4854d016dfada97077fc1696ea23c84e3

File: docs/sources/mimir/manage/mimir-runbooks/_index.md (hunk 15 above)
Comment (ying-jeanne) [reply docs/sources/mimir/manage/mimir-runbooks/_index.md]: flag updated in commit https://github.com/grafana/mimir/commit/1ba1bcb4854d016dfada97077fc1696ea23c84e3

File: docs/sources/mimir/manage/mimir-runbooks/_index.md [hunk 17]
```diff
@@ -2236,6 +2236,20 @@ How to **fix** it:
...
+
+How it **works**:
+
+- The distributor implements an upper limit on the message size of incoming otel write requests after decompression regardless compression type. See more otlp collector [compression details](https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/configgrpc/README.md#client-configuration).
+- To configure the limit, set the `-distributor.max_otlp_request_size`
+
+How to **fix** it:
```
[the remark this answers is not in the document]
Comment (tacole02) [reply docs/sources/mimir/manage/mimir-runbooks/_index.md:2248]: Okay! Thanks for clarifying 😄

File: docs/sources/mimir/manage/mimir-runbooks/_index.md [hunk 18]
```diff
@@ -2236,6 +2236,20 @@ How to **fix** it:
 
 - Increase the allowed limit by using the `-distributor.max-recv-msg-size` option.
 
+### err-mimir-distributor-max-otlp-request-size
+
+This error occurs when a distributor rejects an Otel write request because its message size is larger than the allowed limit after decompression.
+
+How it **works**:
+
+- The distributor implements an upper limit on the message size of incoming OTel write requests after decompression regardless of the compression type. Refer to [OTLP collector compression details](https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/configgrpc/README.md#client-configuration) for more information.
```
Comment (pstibrany) [inline docs/sources/mimir/manage/mimir-runbooks/_index.md:2245]: Referenced documentation seems irrelevant here, because it points to gRPC client config, but Mimir doesn't accept OTLP/gRPC requests.

I don't think we need to mention "compression" at all in this runbook.

File: docs/sources/mimir/manage/mimir-runbooks/_index.md [hunk 19]
```diff
@@ -2236,6 +2236,20 @@ How to **fix** it:
...
+- Configure this limit in the `-distributor.max-otlp-request-size` setting.
+
+How to **fix** it:
+
+- Increase the allowed limit in the `-distributor.max-otlp-request-size` setting.
+- If you use the batch processor in the OTLP collector, decrease the maximum batch size in the `send_batch_max_size` setting. Refer to [Batch Collector](https://github.com/open-telemetry/opentelemetry-collector/blob/main/processor/batchprocessor/README.md) for details.
```
Comment (pstibrany) [inline docs/sources/mimir/manage/mimir-runbooks/_index.md:2251]: I'd suggest to move this as a first suggestion, before increasing the limit. Increasing limit has implications on memory usage.

File: pkg/distributor/push.go [hunk 20]
```diff
@@ -110,6 +110,18 @@ func (e distributorMaxWriteMessageSizeErr) Error() string {
...
+type distributorMaxOTLPRequestSizeErr struct {
+	actual, limit int
+}
+
+func (e distributorMaxOTLPRequestSizeErr) Error() string {
+	msgSizeDesc := fmt.Sprintf(" of %d bytes", e.actual)
+	if e.actual < 0 {
+		msgSizeDesc = ""
+	}
+	return globalerror.DistributorMaxOTLPRequestSize.MessageWithPerInstanceLimitConfig(fmt.Sprintf("the incoming push request has been rejected because its message size%s is larger than the allowed limit of %d bytes", msgSizeDesc, e.limit), maxOTLPRequestSizeFlag)
```
Comment (pstibrany) [inline pkg/distributor/push.go:122]:
```suggestion
	return globalerror.DistributorMaxOTLPRequestSize.MessageWithPerInstanceLimitConfig(fmt.Sprintf("the incoming OTLP request has been rejected because its message size%s is larger than the allowed limit of %d bytes", msgSizeDesc, e.limit), maxOTLPRequestSizeFlag)
```

File: CHANGELOG.md [hunk 21]
```diff
@@ -19,6 +19,7 @@
     * `-ingest-storage.read-consistency`: configures the default read consistency.
     * `-ingest-storage.migration.distributor-send-to-ingesters-enabled`: enabled tee-ing writes to classic ingesters and Kafka, used during a live migration to the new ingest storage architecture.
     * `-ingester.partition-ring.*`: configures partitions ring backend.
+* [ENHANCEMENT] Distributor: Add configuration option `-distributor.max-otlp-request-size` to limit OTel request byte sizes. #8574
```
Comment (pstibrany) [inline CHANGELOG.md:22]:
```suggestion
* [ENHANCEMENT] Distributor: Add configuration option `-distributor.max-otlp-request-size` to limit OTLP request size. #8574
```

File: pkg/distributor/distributor.go [hunk 22]
```diff
@@ -237,6 +240,7 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet, logger log.Logger) {
 	cfg.RetryConfig.RegisterFlags(f)
 
 	f.IntVar(&cfg.MaxRecvMsgSize, "distributor.max-recv-msg-size", 100<<20, "Max message size in bytes that the distributors will accept for incoming push requests to the remote write API. If exceeded, the request will be rejected.")
+	f.IntVar(&cfg.MaxOTLPRequestSize, maxOTLPRequestSizeFlag, 80<<20, "Maximum OTLP request size in bytes that the distributors accept. Requests exceeding this limit are rejected.")
```
Comment (pstibrany) [inline pkg/distributor/distributor.go:243]: How was the 80 MiB selected? Do we have some data about size of the request and processing time?

Intuitively 80 MiB still feels very large. With previous 100 MiB we sometimes see translation to Mimir request taking up to 10 seconds, so perhaps better value would be somewhere close to 10 MiB?

Comment (ying-jeanne) [reply pkg/distributor/distributor.go:243]: as mentioned in the description of the PR, it is tentative. Since 413 error is not retryable, we need to be cautious. if 100MiB/10s, how about 2s of response time ~ 20 MiB

File: docs/sources/mimir/manage/mimir-runbooks/_index.md (hunk 18 above)
Comment (ying-jeanne) [reply docs/sources/mimir/manage/mimir-runbooks/_index.md:2245]: nice catch, gracais

File: pkg/distributor/distributor.go (hunk 22 above)
Comment (pstibrany) [reply pkg/distributor/distributor.go:243]: I agree that we need to be careful when rolling this out. It may be safer to use original value of 100 MiB for now, so that this doesn't break existing setups, and then we can experiment with correct value in our setup.

File: docs/sources/mimir/manage/mimir-runbooks/_index.md (hunk 18 above)
Comment (ying-jeanne) [reply docs/sources/mimir/manage/mimir-runbooks/_index.md]: updated link to confighttp

File: pkg/distributor/distributor.go (hunk 22 above)
Comment (ying-jeanne) [reply pkg/distributor/distributor.go]: let me revert my reverted commit

File: CHANGELOG.md [hunk 23]
```diff
@@ -19,6 +19,7 @@
     * `-ingest-storage.read-consistency`: configures the default read consistency.
     * `-ingest-storage.migration.distributor-send-to-ingesters-enabled`: enabled tee-ing writes to classic ingesters and Kafka, used during a live migration to the new ingest storage architecture.
     * `-ingester.partition-ring.*`: configures partitions ring backend.
+* [CHANGE] Distributor: Previously incoming OTLP requests were size-limited by using limit from `-distributor.max-recv-msg-size` option. We have added option `-distributor.max-otlp-request-size` for OTLP request, which defaults to 100 MiB. #8574
```
Comment (pstibrany) [inline CHANGELOG.md:22]:
```suggestion
* [CHANGE] Distributor: Incoming OTLP requests were previously size-limited by using limit from `-distributor.max-recv-msg-size` option. We have added option `-distributor.max-otlp-request-size` for limiting OTLP requests, with default value of 100 MiB. #8574 
```

File: docs/sources/mimir/configure/about-versioning.md [hunk 24]
```diff
@@ -74,6 +74,8 @@ The following features are currently experimental:
     - `-distributor.retry-after-header.max-backoff-exponent`
   - Limit exemplars per series per request
     - `-distributor.max-exemplars-per-series-per-request`
+  - Limit OTLP write request byte size
+    - `max_otlp_request_size`
```
Comment (pstibrany) [inline docs/sources/mimir/configure/about-versioning.md:78]:
```suggestion
    - `-distributor.max-otlp-request-size`
```

**Review by pstibrany — APPROVED** [review]
Thank you!

**Comment by aknuds1** [comment]
Reviewing again

File: docs/sources/mimir/manage/mimir-runbooks/_index.md [hunk 25]
```diff
@@ -2236,6 +2236,20 @@ How to **fix** it:
 
 - Increase the allowed limit by using the `-distributor.max-recv-msg-size` option.
 
+### err-mimir-distributor-max-otlp-request-size
+
+This error occurs when a distributor rejects an Otel write request because its message size is larger than the allowed limit after decompression.
```
Comment (aknuds1) [inline docs/sources/mimir/manage/mimir-runbooks/_index.md:2241]: The size is checked also prior to decompressing, right:
```suggestion
This error occurs when a distributor rejects an OTel write request because its message size is larger than the allowed limit before or after decompression.
```

File: docs/sources/mimir/manage/mimir-runbooks/_index.md [hunk 26]
```diff
@@ -2236,6 +2236,20 @@ How to **fix** it:
...
+
+How it **works**:
+
+- The distributor implements an upper limit on the message size of incoming OTel write requests after decompression regardless of the compression type. Refer to [OTLP collector compression details](https://github.com/open-telemetry/opentelemetry-collector/tree/main/config/confighttp) for more information.
```
Comment (aknuds1) [inline docs/sources/mimir/manage/mimir-runbooks/_index.md:2245]: The distributor does check the message size also prior to decompressng:
```suggestion
- The distributor implements an upper limit on the message size of incoming OTel write requests before and after decompression regardless of the compression type. Refer to [OTLP collector compression details](https://github.com/open-telemetry/opentelemetry-collector/tree/main/config/confighttp) for more information.
```

**Review by aknuds1 — APPROVED** [review]
LGTM, with a couple of suggested clarifications wrt. compression.

**Comment by ying-jeanne** [comment]
Thank you both! @pstibrany @aknuds1. Let's merge and now set it lower along the time 🤞
